### DISCUSSION TITLE
Enable CURLSSLOPT_NATIVE_CA by default on Windows

### DISCUSSION THREAD

**Discussion by depler** [curl/curl#9348] [Ideas]
Why don't you enable CURLSSLOPT_NATIVE_CA by default so `curl` could use system certificates instead of external file?

**Comment by bagder** [comment]
I don't think it should do it by default as it would change behavior, but I would approve an option to enable it.

**Reply by depler** [reply]
Why not by default? Does it lead to some problems?

**Reply by jay** [reply]
No it should not do it by default. If you build curl with OpenSSL then the usual practice on Windows is that you have a CA bundle or are otherwise using OpenSSL's certificate database and not the Windows certificate database. To change this now would be a breaking change in behavior. If you want to use the Windows CA database store I suggest using curl built with Schannel.

Also, to get in the weeds for a moment, I worked on this option and though it is largely untested (and, also, has no tests) it appeared to mostly work except not very well on CA certificates that share the same details. Windows CA stores over time can amass such certificates. OpenSSL doesn't allow that so if there are any certificates like that it's a toss up which one is actually imported into OpenSSL as the CA certificate, and it could have a different expiration than some other same type cert or different signing algorithm, different cross signing etc. I did my best to filter out some certificates to narrow down the most likely candidates of what OpenSSL would accept but it's almost arbitrary what is actually imported. I just don't know how we would rank the "best" candidate and import that one first.

I'm not against a tool option as long as it carries the same experimental warning. We just can't guarantee or be expected to maintain the issues that may come in from using this option the same as we can with other options. It should mostly work. Mostly...

**Reply by depler** [reply]
Ok, so there are some problems to be expected. Can you at least make an argument switch to enable Windows system certificates from command line? Something like this:

`curl.exe --use-system-certs ...`

**Reply by bagder** [reply]
Someone could do this, sure!

**Comment by vszakats** [comment]
Referencing this Issue, originally opened for curl-for-win: https://github.com/curl/curl/issues/9350

I didn't do tests, but according to the [source](https://github.com/curl/curl/blob/2c86f1b34567b9ba31a3cd3ae4bc2b683d70269c/lib/vtls/openssl.c#L2936) and a quick symbol lookup in `curl.exe`, the OpenSSL native CA feature _is_ present in all Windows builds by default.

As for enabling this behavior by default, I'm with @bagder.
